ftw/crawler/main.py

```diff
@@ -54,77 +55,102 @@ def get_indexing_time(url, indexed_docs, config):
             continue
 
-        # Fetch and parse the sitemap index (or build a virtual one)
-        sitemap_index = SitemapIndexFetcher(site).fetch()
-
-        log.info(u"Crawling {} [{} sitemap(s)]...".format(
-            site.url, len(sitemap_index.sitemaps)))
-
-        # Get all docs indexed in Solr for a particular site
-        indexed_docs = get_indexed_docs(config, solr, site)
-
-        # Purge docs that have been removed from sitemap(s) from Solr index
-        purge_removed_docs_from_index(config, sitemap_index, indexed_docs)
-
-        # Create a requests session to allow for connection pooling
-        fetcher_session = requests.Session()
-
-        for sitemap in sitemap_index.sitemaps:
-            total = len(sitemap.url_infos)
-
-            log.info(u"-" * 78)
-            log.info(u"Crawling {}...".format(sitemap.url))
-
-            for n, url_info in enumerate(sitemap.url_infos, start=1):
-                url = url_info['loc']
-                progress = '[{}/{}]'.format(n, total)
-
-                # If we're only indexing a specific URL, skip all others
-                if options.url and not url == options.url:
-                    continue
-
-                log.debug(u"{}: {}".format(url, unicode(url_info)))
-
-                # Get time this document was last indexed
-                last_indexed = get_indexing_time(url, indexed_docs, config)
-
-                # Fetch and save resource
-                resource_info = ResourceInfo(site=sitemap.site,
-                                             url_info=url_info,
-                                             last_indexed=last_indexed)
-                fetcher = ResourceFetcher(
-                    resource_info, fetcher_session, tempdir, options)
-                try:
-                    resource_info = fetcher.fetch()
-                except NotModified:
-                    log.info(u"{}   Skipped {} (not modified)".format(progress, url))
-                    continue
-                except AttemptedRedirect:
-                    continue
-                except FetchingError, e:
-                    log.error(unicode(e))
-                    continue
-
-                # Extract metadata and plain text
-                engine = ExtractionEngine(
-                    config, resource_info, converter=TikaConverter(config.tika))
-                field_values = engine.extract_field_values()
-                display_fields(field_values)
-                os.unlink(resource_info.filename)
-
-                # Index into Solr
-                log.debug(u"Indexing {} into solr.".format(url))
-                response = solr.index(field_values)
-                if response.status_code == 200:
-                    log.info(u"{} * Indexed {}".format(progress, url))
-
-        log.info(u"=" * 78)
-        log.info(u"")
+        try:
+            crawl_site(tempdir, config, options, solr, site)
+        except KeyboardInterrupt:
```

Comment:
If you were to catch this, you should simply re-raise (just a `raise`) instead of doing `sys.exit()` and suppressing  the stack trace. Particularly not `sys.exit(0)` (exit code for KeyboardInterrupt should be non-zero).

But this shouldn't even be necessary. System-exiting exceptions like `KeyboardInterrupt` and `SystemExit`, and internal-use exceptions like `GeneratorExit` and `StopIteration` [don't inherit from the `Exception` base class any more in Python 2.7](https://docs.python.org/2/library/exceptions.html#exceptions.Exception):

```
>>> issubclass(KeyboardInterrupt, Exception)
False
>>> issubclass(SystemExit, Exception)
False
>>>
>>> issubclass(ValueError, Exception)
True
```